src/coreclr/System.Private.CoreLib/src/System/RuntimeType.ActivatorCache.cs

```diff
@@ -99,32 +126,110 @@ internal ActivatorCache(RuntimeType rt)
                 // method is invoked.
             }
 
-            internal bool CtorIsPublic => _ctorIsPublic;
+            /// <summary>
+            /// Helper to invoke the default (parameterless) constructor.
+            /// </summary>
+            internal object? CreateInstance(RuntimeType rt, bool publicOnly, bool wrapExceptions)
+            {
+                // We don't use RuntimeType, but we force the caller to pass it so
+                // that we can keep it alive on their behalf. Once the object is
+                // constructed, we no longer need the reference to the type instance,
+                // as the object itself will keep the type alive.
 
+#if DEBUG
+                CheckOriginalRuntimeType(rt);
+#endif
+
+                if (_pfnAllocatorForCtorCall == null || (!_ctorIsPublic && publicOnly))
```

Comment:
Is this regressing the regular Activator.CreateInstance performance? Unless you are improving the performance of Activator.CreateInstance path, I do not think you want to touch it.

Reply:
I don't think it should regress, but I will test it. The impl has to change slightly either way because the activator cache needs to be created even when there isn't a default ctor, otherwise it couldn't be used for uninitialized objects with no ctor. It does have one extra `_pfnAllocatorForCtorCall == null` check above, but I don't think that will be noticeable in perf. There is a way to avoid that null check if it's really an issue, by stubbing in an allocator that throws, either for _`pfnAllocatorForCtorCall` or for `_pfnCtor`. 

The code changes are mostly just cleaning it up to work with the new ActivatorCache setup and refactoring out common code that is now shared across 3 methods. I did originally try to keep it how it was but was messy and obtuse with the separated CreateUninitializedInstance and CallConstructor methods.

Reply:
My initial tests indicated 10-15x perf increase for GetUninitializedObject, and aside from that null check, the code for activator is basically the same, just shuffled around a little bit. I wanted to get CI tests going while I finished checking everything else so I started the draft PR.

Reply:
> There is a way to avoid that null check if it's really an issue, by stubbing in a method that throws for _`pfnAllocatorForCtorCall` or `_pfnCtor` if there is no ctor.

Note that this is actually what I had initially, which forced me to shuffle some of the activator code into activator cache to maintain existing activator perf because an extra try/catch is needed to format the exception message correctly, which otherwise would have resulted in an extra non-inlineable method call due to the try/catch. In the end it all seemed like a lot of excess ceremony just to avoid that null check, so I went with this approach and figured I would run some perf tests.